File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        def trade(i,b):
            if  b >= len(prices) or i >= len(prices):
                return 0
            if b == -1:
                return trade(i+1,i)
            t = 0
            if prices[i] > prices[b]:
                t = prices[i] - prices[b]
            return max(trade(i+1,b),t + trade(i+2,-1))
        m = 0
        for i in range(len(prices)):
            m = max(m,trade(i,-1))
        return m

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<int> suffix(prices.size());
        int profit = 0;
        int t= 0;
        for(int i = prices.size()-1;i>=0;i--){
            suffix[i] = t;
            if (prices[i] > t){
                t = prices[i];
            }
        }

        for (int i =0;i<prices.size();i++){
            profit = max(profit , suffix[i]-prices[i]);
        }

        return profit;
    }
};
